quantum/audio/audio_chibios_pwm.c

```diff
@@ -0,0 +1,195 @@
+#else // AUDIO_DRIVER_PWM
+                           .callback = pwm_audio_period_callback,
+#endif
+                           .channels = {
+#if AUDIO_DRIVER_PWM_PIN_ALTERNATE
+                               {PWM_OUTPUT_ACTIVE_HIGH, NULL}, /* channel 0 -> TIM1_CH1 */
+#else // AUDIO_DRIVER_PWM
+                               {PWM_OUTPUT_ACTIVE_HIGH, pwm_audio_channel_interrupt_callback}, /* channel 0 -> TIM1_CH1 */
+#endif
+                               {PWM_OUTPUT_DISABLED, NULL}, /* channel 1 -> TIM1_CH2 */
+                               {PWM_OUTPUT_DISABLED, NULL}, /* channel 2 -> TIM1_CH3 */
+                               {PWM_OUTPUT_DISABLED, NULL}  /* channel 3 -> TIM1_CH4 */
+                           },
+                           };
+
+static float channel_1_frequency = 0.0f;
+void channel_1_set_frequency(float freq) {
+    if (freq == channel_1_frequency) return;
+
+    channel_1_frequency = freq;
+
+    pwmcnt_t period = (pwmCFG.frequency / freq);
+    pwmChangePeriod(&PWMD1, period);
+    pwmEnableChannel(&PWMD1, 0, PWM_PERCENTAGE_TO_WIDTH(&PWMD1, 5000));
+}
+
+float channel_1_get_frequency(void) { return channel_1_frequency; }
+
+void channel_1_start(void) {
+    pwmStop(&PWMD1);
+    pwmStart(&PWMD1, &pwmCFG);
+
+#if defined(AUDIO_DRIVER_PWM)
+    pwmEnablePeriodicNotification(&PWMD1);
+    pwmEnableChannelNotification(&PWMD1, 0);
+#endif
+}
+
+void channel_1_stop(void) {
+    pwmStop(&PWMD1);
+
+#if defined(AUDIO_DRIVER_PWM)
+    palClearLine(AUDIO_PIN); // leave the line low, after last note was played
+#endif
+}
+
+#if defined(AUDIO_DRIVER_PWM)
+// generate a PWM signal on any pin, not neccessarily the one connected to the timer
+static void pwm_audio_period_callback(PWMDriver *pwmp) {
+    (void)pwmp;
+    palClearLine(AUDIO_PIN);
+}
+static void pwm_audio_channel_interrupt_callback(PWMDriver *pwmp) {
+    (void)pwmp;
+    palSetLine(AUDIO_PIN);  // generate a PWM signal on any pin, not neccessarily the one connected to the timer
+}
+#endif // AUDIO_DRIVER_PWM
+
+static void gpt_callback(GPTDriver *gptp);
+GPTConfig   gptCFG = {
+    /* a whole note is one beat, which is - per definition in musical_notes.h - set to 64 
+       the longest note is BREAVE_DOT=128+64=192, the shortest SIXTEENTH=4
+       the tempo (which might vary!) is in bpm (beats per minute)
+       therefore: if the timer ticks away at .frequency = (60*64)Hz,
+       and the .intervall counts from 64 downwards - all we need to do is increment the
+       note_position on each callback, and have the note_lengt = duration*tempo compare
+       against that; hence: audio_advance_note(step=1, end=1)
+    */
+    .frequency = 60 * 64,
+    .callback  = gpt_callback,
+};
+
+void audio_initialize_hardware(void) {
+    pwmStart(&PWMD1, &pwmCFG);
+
+#if defined(AUDIO_DRIVER_PWM_PIN_ALTERNATE)
+#    if defined(PAL_MODE_STM32_ALTERNATE_PUSHPULL)
+    //TODO: is there a better way to differentiate between chibios GPIOv1 and GPIOv2?
```

Comment:
We've started with `tmk_core/common/chibios/chibios_config.h`. If you need to put extra definitions to determine specifics, it'd be great if it could go there.
We already have a `USE_GPIOV1` present, at any rate.